Use Kustomize to customize the service installation (#512)
akka/akka-platform-guide: docs-source/docs/modules/deployment/pages/operator-reference.adoc

A fictive example of a https://lightbend.github.io/akka-operator-helm/api-reference/v1/shopping-cart-service-full-cr.yml[deployment descriptor (CR) containing all properties {tab-icon}, window="tab"].

> To apply customizations on top of a default deployment descriptor (CR) we suggest to leverage https://kustomize.io[Kustomize] built into `kubectl`.

Reviewer: ```suggestion
To apply customizations on top of a default deployment descriptor (CR) we recommend using https://kubectl.docs.kubernetes.io/guides/introduction/kustomize/[Kustomize], built into `kubectl`.
```

> To apply customizations on top of a default deployment descriptor (CR) we suggest to leverage https://kustomize.io[Kustomize] built into `kubectl`.

Reviewer 2: I suggest that we describe the kustomize approach in a separate section and keep the examples here as they were. I see kustomize as an optional approach.

It can be a separate `==` section in this `operator-reference.adoc`. At some point I think we should extract it into a [how-to guide instead](https://akka.io/platform-guide/wip/how-to/index.html) but it's ok to keep all these additional topics in operator-reference.adoc for now.

Author: My personal bias is that the kustomize approach is arguably cleaner, more "production-ready" and helps the user to follow the most recent kubernetes best practices.
Sticking with `kubectl apply -f` we decide to favor retro-compatibility and a more bare-bone "demo" approach.

But having to keep and maintain the 2 different approaches seems like unnecessary work and maintenance.
In case we prefer `kubectl apply -f` I'm ok with dropping this PR.

Reviewer 3: +1 on kustomize approach

Reviewer 2: I want the examples to be as simple as possible, introducing as few new tools as possible. We can revisit that later but I want us to keep it simple to begin with.

However, I think it's great to describe the possibility.

Reviewer 3: It's important to note that kustomize is not another tool. It is a feature of kubectl that is just there.

Author: > I want the examples to be as simple as possible

Having less code copy-pasted makes them easier and nicer to follow IMO.

> introducing as few new tools as possible

kustomize is embedded in kubectl, we are not introducing new tools.
We are, as well, following the lead of kubernetes official documentation e.g.:
https://kubectl.docs.kubernetes.io/guides/config_management/secrets_configmaps/#configmapsgenerator
https://kubectl.docs.kubernetes.io/guides/app_deployment/publishing_bases/#publishing-a-white-box-base

Author: @patriknw for context I have opened #515 based on this approach.
Having to describe the process without kustomize is going to be much more awkward and involved.

Reviewer 2: > kustomize is embedded in kubectl, we are not introducing new tools.

I know, but in my opinion it's adding additional complexity if we would use it in all tutorial examples. More difficult to understand how things are merged together than just one yml file.

Author: We are going in loops, I do understand where you come from @patriknw but I do not share the same opinion.

In this case, I read `additional complexity` as `separation of concerns` that is actually desirable.

I have the impression that @RayRoestenburg and @seglo share my same opinion, but, let's agree to disagree and move on.
If we don't take this approach I need more guidance on how you would structure and phrase i.e. #515 to be able to follow up (i.e. do you want to include instructions on how to base64 the secret? ...).
Thanks!

Reviewer 2: what do you mean with base64? secret can be created with
```
kubectl create secret generic shopping-cart-service-astra-secret --from-literal=username=scott --from-literal=password=tiger --from-file=secureConnectBundle=/path/to/secure-connect.zip
```

Reviewer 2: > In this case, I read additional complexity as separation of concerns that is actually desirable.

I'm talking about the tutorial example: https://akka.io/platform-guide/wip/microservices-tutorial/grpc-service.html#kubernetes

The additional kuztomize file and mandatory directory structure would introduce complexity to that example, without immediate value for those that are learning this for the first time.

Reviewer 4: I agree that kustomize makes it harder for users new to Kubernetes and those users that haven't adopted it; requiring a user to understand resources and that they can be composed from multiple files is more to grasp than here's everything in one file, go and apply it to your cluster.

I don't want us to have to explain kustomize in this guide or for a user to have to go and learn what it is. Best practice defined by blogs/conferences doesn't mean the majority of k8s users use it yet. The separation of concerns makes it simpler, but not easier. 

That said the point of the operator, IMO, is to make deploying Akka easier with fewer things required in resources as we have sensible defaults / manage many of resources for you. So I don't think it matters which approach we show in the guide so we should decide quickly and move on.

Reviewer 3: Kustomize provides an easy to learn way to organize the resource you want to apply, including secrets that you refer to. 
No everything is defined in one file, not even right now.

> Best practice defined by blogs/conferences doesn't mean the majority of k8s users use it yet.
Hm.... An assumption that can go either way. We're not advocating for it because we read it on a blog obviously, but because it is an easy way to structure more than one resource in a convenient way, and many have used this on the job. Many kubectl invocations, for the user to copy paste is not very useful.

Reviewer 3: I would say use kustomize in the examples where it makes sense. There is no point in kustomize if there is nothing to customize, very obviously. But where it is useful we should not be scared of it. The user is already supposedly fine with many kubectl commands. They will copy / paste whatever we give them right?

Reviewer 3: Have we thought of supporting distinct environments? dev / stage / prod / etc. Users will likely need that once they stop playing around, just copy/pasting commands from our docs. This can be nicely structured with kustomize, so getting the user accustomed to that is not a bad thing?

Reviewer 2: > I would say use kustomize in the examples where it makes sense. There is no point in kustomize if there is nothing to customize, very obviously. But where it is useful we should not be scared of it.

That's what I have been trying to say.  I suggested that we describe the kustomize usage in a separate how-to section (or separate section in this operator-reference.adoc). Leaving the examples in the tutorial as plain cr-yml files and kubectl commands for creating secrets.

> Have we thought of supporting distinct environments? dev / stage / prod / etc. Users will likely need that once they stop playing around, just copy/pasting commands from our docs.

That can also be part of that kustomize how-to. Later, I'm pretty sure we will have stories around our recommendations for CI/CD and then that might play an important role (or not). We are not at that working on that story yet (well the story might not even exist yet).

Prepare a folder called `base` containing the following files:

The default deployment descriptor (CR):
[source,yaml,title='shopping-cart-service.yml']
----
apiVersion: akka.lightbend.com/v1
kind: AkkaMicroservice
metadata:
  name: shopping-cart-service
spec:
  image: <image>
----

A basic reference to it for Kustomize:
[source,yaml,title='kustomization.yml']
----
resources:
- shopping-cart-service.yml
----

> You can already deploy you default service by running:

Reviewer: ```suggestion
Deploy your default service by running:
```

[source,shell script]
----
kubectl apply -k base
----

== Scale the number of Pods

The `replicas` field in the deployment descriptor defines the number of Pods that will be used.

Create a separate folder along `base` and call it `replicas` with this content:

The actual customization:
[source,yaml,title='replicas.yml']
----
apiVersion: akka.lightbend.com/v1
kind: AkkaMicroservice
metadata:
  name: shopping-cart-service
spec:
  replicas: 3
----

And a reference to the customization:
[source,yaml,title='kustomization.yml']
----
bases:
- ../base
patchesStrategicMerge:
- replicas.yaml
----

> You can scale the number of pods by applying the changed deployment descriptor by running:

Reviewer: ```suggestion
Scale the number of pods by applying the overlaid deployment descriptor by running:
```

[source,shell script]
----
kubectl apply -k replicas
----

> Kubernetes will add or remove Pods to adjust to the desired number.

Reviewer: ```suggestion
Kubernetes will reconcile the number of Pods to the desired scale.
```
